src/transformers/file_utils.py

```diff
@@ -1943,41 +1948,48 @@ def push_to_hub(
         Returns:
             The url of the commit of your model in the given repository.
         """
-        with tempfile.TemporaryDirectory() as tmp_dir:
-            self.save_pretrained(tmp_dir)
-            self._push_to_hub(
-                save_directory=tmp_dir,
-                repo_name=repo_name,
-                repo_url=repo_url,
-                commit_message=commit_message,
-                organization=organization,
-                private=private,
-                use_auth_token=use_auth_token,
-            )
+        working_dir = working_directory if working_directory is not None else tempfile.tempfile.mkdtemp()
+        # Create or clone the repo
+        repo = self._create_or_get_repo(
+            working_directory=working_dir,
+            repo_name=repo_name,
+            repo_url=repo_url,
+            organization=organization,
+            private=private,
+            use_auth_token=use_auth_token,
+        )
+        # Save the files in the cloned repo
+        self.save_pretrained(working_dir)
+        # Commit and push!
+        url = self._push_to_hub(repo, commit_message=commit_message)
 
-    @classmethod
-    def _push_to_hub(
-        cls,
-        save_directory: Optional[str] = None,
-        save_files: Optional[List[str]] = None,
+        # Clean up! Clean up! Everybody everywhere!
+        if working_directory is None:
+            shutil.rmtree(working_dir)
+
+        return url
+
+    def _create_or_get_repo(
+        self,
+        working_directory: Optional[str] = None,
         repo_name: Optional[str] = None,
         repo_url: Optional[str] = None,
-        commit_message: Optional[str] = None,
         organization: Optional[str] = None,
         private: bool = None,
         use_auth_token: Optional[Union[bool, str]] = None,
-    ) -> str:
-        # Private version of push_to_hub, that either accepts a folder to push or a list of files.
-        if save_directory is None and save_files is None:
-            raise ValueError("_push_to_hub requires either a `save_directory` or a list of `save_files`.")
-        if repo_name is None and repo_url is None and save_directory is None:
+    ):
+        if repo_name is None and repo_url is None and working_directory is None:
             raise ValueError("Need either a `repo_name` or `repo_url` to know where to push!")
 
-        if repo_name is None and repo_url is None and save_files is None:
-            repo_name = Path(save_directory).name
+        if repo_name is None and repo_url is None:
+            repo_name = Path(working_directory).name
```

Comment:
i'd say you either need to pass a repo_name or a repo_url

Reply:
If a the repo cloned locally and want to update the config for instance, this would have me do
```
config.push_to_hub(repo_name="my-awesome-model", working_directory="~/git/models/my-awesome-model")
```
when we can get all the info from:
```
config.push_to_hub(working_directory="~/git/models/my-awesome-model")
```